csharp/src/Apache.Arrow/Ipc/ArrowFileReaderImplementation.cs

```diff
@@ -70,36 +72,79 @@ protected override async Task ReadSchemaAsync()
             {
                 BaseStream.Position = BaseStream.Length - ArrowFileConstants.Magic.Length - 4;
 
-                bytesRead = await BaseStream.ReadAsync(buffer, 0, 4).ConfigureAwait(false);
-                footerLength = BinaryPrimitives.ReadInt32LittleEndian(buffer);
+                bytesRead = await BaseStream.ReadAsync(buffer.AsMemory(0, 4)).ConfigureAwait(false);
+                footerLength = ReadFooterLength(buffer, bytesRead);
+            }).ConfigureAwait(false);
 
-                if (bytesRead != 4) throw new InvalidDataException(
-                    $"Failed to read footer length. Read <{bytesRead}>, expected 4.");
+            await Buffers.RentReturnAsync(footerLength, async (buffer) =>
+            {
+                _footerStartPostion = (int)BaseStream.Length - footerLength - ArrowFileConstants.Magic.Length - 4;
 
-                if (footerLength <= 0) throw new InvalidDataException(
-                    $"Footer length has invalid size <{footerLength}>");
+                BaseStream.Position = _footerStartPostion;
+
+                bytesRead = await BaseStream.ReadAsync(buffer.AsMemory(0, footerLength)).ConfigureAwait(false);
+                ReadSchema(buffer, bytesRead, footerLength);
             }).ConfigureAwait(false);
+        }
 
-            await Buffers.RentReturnAsync(footerLength, async (buffer) =>
+        protected override void ReadSchema()
+        {
+            if (HasReadSchema)
+            {
+                return;
+            }
+
+            ValidateFile();
+
+            var bytesRead = 0;
+            var footerLength = 0;
+
+            Buffers.RentReturn(4, (buffer) =>
+            {
+                BaseStream.Position = BaseStream.Length - ArrowFileConstants.Magic.Length - 4;
```

Comment:
`BaseStream.Length - ArrowFileConstants.Magic.Length - 4` appears several times, could it be extracted into a well named method?

Reply:
This was fixed in #4017.